src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3645,74 +3693,114 @@ protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOper
                 binary = stack.Pop();
             }
 
-            // learn from true/false constant
-            bool learnFromBooleanConstantTest()
+            static bool isEqualsOrNotEquals(BoundBinaryOperator binary)
+                => binary.OperatorKind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual;
+
+            static bool isEquals(BoundBinaryOperator binary)
+                => binary.OperatorKind.Operator() == BinaryOperatorKind.Equal;
+
+            static bool isKnownNullOrNotNull(BoundExpression expr, TypeWithState resultType)
+            {
+                return resultType.State.IsNotNull()
+                    || expr.ConstantValue is object;
+            }
+
+            LocalState getSingleState(BoundExpression otherOperand, PossiblyConditionalState conditionalStateWhenNotNull)
```

Comment:
>getSingleState [](http://example.com/codeflow?start=23&length=14)

nit: `getUnconditionalState`?

Reply:
Have changed the name to try and improve clarity